File: fs/ArchSims.Core/Ahmes.fs
namespace Ufrgs.Inf.ArchSims.Core

open LanguagePrimitives

open Ufrgs.Inf.ArchSims.Core.Memory

module Ahmes =

    type Instruction =
    | Nop = 0x00uy // 0000 0000
    | Sta = 0x10uy // 0001 0000
    | Lda = 0x20uy // 0010 0000
    | Add = 0x30uy // 0011 0000
    | Or  = 0x40uy // 0100 0000
    | And = 0x50uy // 0101 0000
    | Not = 0x60uy // 0110 0000
    | Sub = 0x70uy // 0111 0000
    | Jmp = 0x80uy // 1000 0000
    | Jn  = 0x90uy // 1001 0000
    | Jp  = 0x94uy // 1001 0100
    | Jv  = 0x98uy // 1001 1000
    | Jnv = 0x9Cuy // 1001 1100
    | Jz  = 0xA0uy // 1010 0000
    | Jnz = 0xA4uy // 1010 0100
    | Jc  = 0xB0uy // 1011 0000
    | Jnc = 0xB4uy // 1011 0100
    | Jb  = 0xB8uy // 1011 1000
    | Jnb = 0xBCuy // 1011 1100
    | Shr = 0xE0uy // 1110 0000
    | Shl = 0xE1uy // 1110 0001
    | Ror = 0xE2uy // 1110 0010
    | Rol = 0xE3uy // 1110 0011
    | Hlt = 0xF0uy // 1111 0000

    type InstructionRegister = {
        mutable OpCode: byte
        mutable OperandAddress: byte
    }

    type Flags = {
        mutable Halted: bool
        mutable Negative: bool
        mutable Zero: bool
        mutable Carry: bool
        mutable Overflow: bool
        mutable Borrow: bool
    }

    type Registers = {
        mutable ProgramCounter: byte
        mutable Accumulator: byte

        InstructionRegister: InstructionRegister
        Flags: Flags
    }

    let CreateRegisters() = {
        ProgramCounter = 0uy
        Accumulator = 0uy
        InstructionRegister = { OpCode = 0uy; OperandAddress = 0uy }
        Flags = { Halted = false; Negative = false; Zero = true; Carry = false; Overflow = false; Borrow = false }
    }

    let RegistersReset registers =
        registers.ProgramCounter <- 0uy
        registers.Accumulator <- 0uy
        registers.InstructionRegister.OpCode <- 0uy
        registers.InstructionRegister.OperandAddress <- 0uy
        registers.Flags.Halted <- false
        registers.Flags.Negative <- false
        registers.Flags.Zero <- true
        registers.Flags.Carry <- false
        registers.Flags.Overflow <- false
        registers.Flags.Borrow <- false

    type Cpu = {
        Registers: Registers
        Memory: Memory
    }

    let CreateCpu() = {
        Registers = CreateRegisters()
        Memory = CreateMemory 256
    }

    let Fetch cpu =
        let readByteFromProgramCounterAndAdvance() =
            let result = MemoryReadByte cpu.Memory (int cpu.Registers.ProgramCounter)
            cpu.Registers.ProgramCounter <- cpu.Registers.ProgramCounter + 1uy
            result

        cpu.Registers.InstructionRegister.OpCode <- readByteFromProgramCounterAndAdvance()
        cpu.Registers.InstructionRegister.OperandAddress <-
            match EnumOfValue cpu.Registers.InstructionRegister.OpCode with
            | Instruction.Sta
            | Instruction.Lda
            | Instruction.Add
            | Instruction.Or
            | Instruction.And
            | Instruction.Sub
            | Instruction.Jmp
            | Instruction.Jn
            | Instruction.Jp
            | Instruction.Jv
            | Instruction.Jnv
            | Instruction.Jz
            | Instruction.Jnz
            | Instruction.Jc
            | Instruction.Jnc
            | Instruction.Jb
            | Instruction.Jnb -> readByteFromProgramCounterAndAdvance()  // Instructions with operand
            | _ -> 0uy // Instructions without operand

    let Execute cpu =
        let readOperand() =
            MemoryReadByte cpu.Memory (int cpu.Registers.InstructionRegister.OperandAddress)

        let isNegativeByte (value: byte) = value > 0x7Fuy

        let writeAccumulator value =
            cpu.Registers.Accumulator <- value
            cpu.Registers.Flags.Zero <- value = 0uy
            cpu.Registers.Flags.Negative <- isNegativeByte value

        let addAndUpdateFlags (a: byte) (b: byte) =
            let full = int a + int b
            cpu.Registers.Flags.Carry <- full > 0xFF
            let result = byte full
            let signA = isNegativeByte a
            let signB = isNegativeByte b
            let signR = isNegativeByte result
            cpu.Registers.Flags.Overflow <- (signA = signB) && (signA <> signR)
            result

        let subAndUpdateFlags (a: byte) (b: byte) =
            let full = int a - int b
            cpu.Registers.Flags.Borrow <- full < 0
            let result = byte (full &&& 0xFF)
            let signA = isNegativeByte a
            let signB = isNegativeByte b
            let signR = isNegativeByte result
            cpu.Registers.Flags.Overflow <- (signA <> signB) && (signA <> signR)
            result

        let jumpIf condition =
            if condition then
                cpu.Registers.ProgramCounter <- cpu.Registers.InstructionRegister.OperandAddress

        let instruction = EnumOfValue cpu.Registers.InstructionRegister.OpCode
        match instruction with
        | Instruction.Sta ->   // STA addr: MEM(addr) <- AC
            cpu.Registers.Accumulator |> MemoryWriteByte cpu.Memory (int cpu.Registers.InstructionRegister.OperandAddress)

        | Instruction.Lda ->   // LDA addr: AC <- MEM(addr)
            readOperand() |> writeAccumulator

        | Instruction.Add ->   // ADD addr: AC <- AC + MEM(addr)
            addAndUpdateFlags cpu.Registers.Accumulator (readOperand()) |> writeAccumulator

        | Instruction.Or ->    // OR addr: AC <- AC or MEM(addr)
            cpu.Registers.Accumulator ||| readOperand() |> writeAccumulator

        | Instruction.And ->   // AND addr: AC <- AC and MEM(addr)
            cpu.Registers.Accumulator &&& readOperand() |> writeAccumulator

        | Instruction.Not ->   // NOT: AC <- not AC
            ~~~cpu.Registers.Accumulator |> writeAccumulator

        | Instruction.Sub ->   // SUB addr: AC <- AC - MEM(addr)
            subAndUpdateFlags cpu.Registers.Accumulator (readOperand()) |> writeAccumulator

        | Instruction.Jmp -> jumpIf true                                            // JMP addr
        | Instruction.Jn  -> jumpIf cpu.Registers.Flags.Negative                    // JN  addr: IF N=1
        | Instruction.Jp  -> jumpIf (not cpu.Registers.Flags.Negative)              // JP  addr: IF N=0
        | Instruction.Jv  -> jumpIf cpu.Registers.Flags.Overflow                    // JV  addr: IF V=1
        | Instruction.Jnv -> jumpIf (not cpu.Registers.Flags.Overflow)              // JNV addr: IF V=0
        | Instruction.Jz  -> jumpIf cpu.Registers.Flags.Zero                        // JZ  addr: IF Z=1
        | Instruction.Jnz -> jumpIf (not cpu.Registers.Flags.Zero)                  // JNZ addr: IF Z=0
        | Instruction.Jc  -> jumpIf cpu.Registers.Flags.Carry                       // JC  addr: IF C=1
        | Instruction.Jnc -> jumpIf (not cpu.Registers.Flags.Carry)                 // JNC addr: IF C=0
        | Instruction.Jb  -> jumpIf cpu.Registers.Flags.Borrow                      // JB  addr: IF B=1
        | Instruction.Jnb -> jumpIf (not cpu.Registers.Flags.Borrow)                // JNB addr: IF B=0

        | Instruction.Shr ->   // SHR: Logical shift right; C <- AC[0]; AC <- AC >> 1; AC[7] <- 0
            let ac = cpu.Registers.Accumulator
            cpu.Registers.Flags.Carry <- (ac &&& 0x01uy) <> 0uy
            writeAccumulator (ac >>> 1)

        | Instruction.Shl ->   // SHL: Logical shift left; C <- AC[7]; AC <- AC << 1; AC[0] <- 0
            let ac = cpu.Registers.Accumulator
            cpu.Registers.Flags.Carry <- (ac &&& 0x80uy) <> 0uy
            writeAccumulator (ac <<< 1)

        | Instruction.Ror ->   // ROR: Rotate right through carry
            let ac = cpu.Registers.Accumulator
            let newHighBit = if cpu.Registers.Flags.Carry then 0x80uy else 0x00uy
            cpu.Registers.Flags.Carry <- (ac &&& 0x01uy) <> 0uy
            writeAccumulator ((ac >>> 1) ||| newHighBit)

        | Instruction.Rol ->   // ROL: Rotate left through carry
            let ac = cpu.Registers.Accumulator
            let newLowBit = if cpu.Registers.Flags.Carry then 0x01uy else 0x00uy
            cpu.Registers.Flags.Carry <- (ac &&& 0x80uy) <> 0uy
            writeAccumulator ((ac <<< 1) ||| newLowBit)

        | Instruction.Hlt      // HLT, NOP or unknown instruction: nothing to do
        | Instruction.Nop
        | _ -> ()

        cpu.Registers.Flags.Halted <- instruction = Instruction.Hlt

    let Step cpu =
        Fetch cpu
        Execute cpu

    let Reset cpu =
        cpu.Registers |> RegistersReset
        cpu.Memory |> MemoryReset

File: web/src/core/ahmes.ts
import {
  createMemory,
  memoryReadByte,
  memoryReset,
  memoryWriteByte,
  type Memory,
} from './memory';

export const AhmesInstruction = {
  Nop: 0x00,
  Sta: 0x10,
  Lda: 0x20,
  Add: 0x30,
  Or: 0x40,
  And: 0x50,
  Not: 0x60,
  Sub: 0x70,
  Jmp: 0x80,
  Jn: 0x90,
  Jp: 0x94,
  Jv: 0x98,
  Jnv: 0x9c,
  Jz: 0xa0,
  Jnz: 0xa4,
  Jc: 0xb0,
  Jnc: 0xb4,
  Jb: 0xb8,
  Jnb: 0xbc,
  Shr: 0xe0,
  Shl: 0xe1,
  Ror: 0xe2,
  Rol: 0xe3,
  Hlt: 0xf0,
} as const;

export type AhmesInstruction = (typeof AhmesInstruction)[keyof typeof AhmesInstruction];

export interface AhmesInstructionRegister {
  opCode: number;
  operandAddress: number;
}

export interface AhmesFlags {
  halted: boolean;
  negative: boolean;
  zero: boolean;
  carry: boolean;
  overflow: boolean;
  borrow: boolean;
}

export interface AhmesRegisters {
  programCounter: number;
  accumulator: number;
  instructionRegister: AhmesInstructionRegister;
  flags: AhmesFlags;
}

export interface AhmesCpu {
  registers: AhmesRegisters;
  memory: Memory;
}

export function createAhmesRegisters(): AhmesRegisters {
  return {
    programCounter: 0,
    accumulator: 0,
    instructionRegister: { opCode: 0, operandAddress: 0 },
    flags: {
      halted: false,
      negative: false,
      zero: true,
      carry: false,
      overflow: false,
      borrow: false,
    },
  };
}

export function ahmesRegistersReset(registers: AhmesRegisters): void {
  registers.programCounter = 0;
  registers.accumulator = 0;
  registers.instructionRegister.opCode = 0;
  registers.instructionRegister.operandAddress = 0;
  registers.flags.halted = false;
  registers.flags.negative = false;
  registers.flags.zero = true;
  registers.flags.carry = false;
  registers.flags.overflow = false;
  registers.flags.borrow = false;
}

export function createAhmesCpu(): AhmesCpu {
  return { registers: createAhmesRegisters(), memory: createMemory(256) };
}

function instructionTakesOperand(opCode: number): boolean {
  switch (opCode) {
    case AhmesInstruction.Sta:
    case AhmesInstruction.Lda:
    case AhmesInstruction.Add:
    case AhmesInstruction.Or:
    case AhmesInstruction.And:
    case AhmesInstruction.Sub:
    case AhmesInstruction.Jmp:
    case AhmesInstruction.Jn:
    case AhmesInstruction.Jp:
    case AhmesInstruction.Jv:
    case AhmesInstruction.Jnv:
    case AhmesInstruction.Jz:
    case AhmesInstruction.Jnz:
    case AhmesInstruction.Jc:
    case AhmesInstruction.Jnc:
    case AhmesInstruction.Jb:
    case AhmesInstruction.Jnb:
      return true;
    default:
      return false;
  }
}

export function ahmesFetch(cpu: AhmesCpu): void {
  const advance = (): number => {
    const value = memoryReadByte(cpu.memory, cpu.registers.programCounter);
    cpu.registers.programCounter = (cpu.registers.programCounter + 1) & 0xff;
    return value;
  };
  const ir = cpu.registers.instructionRegister;
  ir.opCode = advance();
  ir.operandAddress = instructionTakesOperand(ir.opCode) ? advance() : 0;
}

const isNegativeByte = (value: number): boolean => (value & 0xff) > 0x7f;

export function ahmesExecute(cpu: AhmesCpu): void {
  const r = cpu.registers;
  const ir = r.instructionRegister;
  const op = ir.opCode;
  const flags = r.flags;
  const readOperand = (): number => memoryReadByte(cpu.memory, ir.operandAddress);

  const writeAc = (value: number): void => {
    const v = value & 0xff;
    r.accumulator = v;
    flags.zero = v === 0;
    flags.negative = v > 0x7f;
  };

  const addAndUpdateFlags = (a: number, b: number): number => {
    const full = (a & 0xff) + (b & 0xff);
    flags.carry = full > 0xff;
    const result = full & 0xff;
    const sa = isNegativeByte(a);
    const sb = isNegativeByte(b);
    const sr = isNegativeByte(result);
    flags.overflow = sa === sb && sa !== sr;
    return result;
  };

  const subAndUpdateFlags = (a: number, b: number): number => {
    const full = (a & 0xff) - (b & 0xff);
    flags.borrow = full < 0;
    const result = full & 0xff;
    const sa = isNegativeByte(a);
    const sb = isNegativeByte(b);
    const sr = isNegativeByte(result);
    flags.overflow = sa !== sb && sa !== sr;
    return result;
  };

  const jumpIf = (cond: boolean): void => {
    if (cond) r.programCounter = ir.operandAddress;
  };

  switch (op) {
    case AhmesInstruction.Sta:
      memoryWriteByte(cpu.memory, ir.operandAddress, r.accumulator);
      break;
    case AhmesInstruction.Lda:
      writeAc(readOperand());
      break;
    case AhmesInstruction.Add:
      writeAc(addAndUpdateFlags(r.accumulator, readOperand()));
      break;
    case AhmesInstruction.Or:
      writeAc(r.accumulator | readOperand());
      break;
    case AhmesInstruction.And:
      writeAc(r.accumulator & readOperand());
      break;
    case AhmesInstruction.Not:
      writeAc(~r.accumulator);
      break;
    case AhmesInstruction.Sub:
      writeAc(subAndUpdateFlags(r.accumulator, readOperand()));
      break;
    case AhmesInstruction.Jmp:
      jumpIf(true);
      break;
    case AhmesInstruction.Jn:
      jumpIf(flags.negative);
      break;
    case AhmesInstruction.Jp:
      jumpIf(!flags.negative);
      break;
    case AhmesInstruction.Jv:
      jumpIf(flags.overflow);
      break;
    case AhmesInstruction.Jnv:
      jumpIf(!flags.overflow);
      break;
    case AhmesInstruction.Jz:
      jumpIf(flags.zero);
      break;
    case AhmesInstruction.Jnz:
      jumpIf(!flags.zero);
      break;
    case AhmesInstruction.Jc:
      jumpIf(flags.carry);
      break;
    case AhmesInstruction.Jnc:
      jumpIf(!flags.carry);
      break;
    case AhmesInstruction.Jb:
      jumpIf(flags.borrow);
      break;
    case AhmesInstruction.Jnb:
      jumpIf(!flags.borrow);
      break;
    case AhmesInstruction.Shr: {
      const ac = r.accumulator;
      flags.carry = (ac & 0x01) !== 0;
      writeAc(ac >>> 1);
      break;
    }
    case AhmesInstruction.Shl: {
      const ac = r.accumulator;
      flags.carry = (ac & 0x80) !== 0;
      writeAc(ac << 1);
      break;
    }
    case AhmesInstruction.Ror: {
      const ac = r.accumulator;
      const newHighBit = flags.carry ? 0x80 : 0x00;
      flags.carry = (ac & 0x01) !== 0;
      writeAc((ac >>> 1) | newHighBit);
      break;
    }
    case AhmesInstruction.Rol: {
      const ac = r.accumulator;
      const newLowBit = flags.carry ? 0x01 : 0x00;
      flags.carry = (ac & 0x80) !== 0;
      writeAc((ac << 1) | newLowBit);
      break;
    }
    default:
      // Hlt, Nop, or unknown opcode: nothing else to do.
      break;
  }

  flags.halted = op === AhmesInstruction.Hlt;
}

export function ahmesStep(cpu: AhmesCpu): void {
  ahmesFetch(cpu);
  ahmesExecute(cpu);
}

export function ahmesReset(cpu: AhmesCpu): void {
  ahmesRegistersReset(cpu.registers);
  memoryReset(cpu.memory);
}
